src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp

```diff
@@ -104,14 +112,27 @@ void ShenandoahMark::do_task(ShenandoahObjToScanQueue* q, T* cl, ShenandoahLiveD
 }
 
 template <ShenandoahGenerationType GENERATION>
-inline void ShenandoahMark::count_liveness(ShenandoahLiveData* live_data, oop obj) {
-  ShenandoahHeap* const heap = ShenandoahHeap::heap();
-  size_t region_idx = heap->heap_region_index_containing(obj);
-  ShenandoahHeapRegion* region = heap->get_region(region_idx);
-  size_t size = obj->size();
+inline void ShenandoahMark::count_liveness(ShenandoahLiveData* live_data, oop obj, uint worker_id) {
+  const ShenandoahHeap* const heap = ShenandoahHeap::heap();
+  const size_t region_idx = heap->heap_region_index_containing(obj);
+  ShenandoahHeapRegion* const region = heap->get_region(region_idx);
+  const size_t size = obj->size();
+
+  // Age census for objects in the young generation
+  if (GENERATION == YOUNG || (GENERATION == GLOBAL && region->is_young())) {
```

Comment:
If I read this correctly, this condition will generate a runtime check region->is_young() in single-gen Shenandoah. 1. This seems like something that would never happen (in single-gen GC) and 2. This path is critical for performance of concurrent marking. Please make sure to keep it as fast as possible, also with respect to the following flag-checks.

Reply:
There is a separate `ShenandoahGenerationType` for non-generational modes: `NON_GEN`. https://github.com/openjdk/jdk/pull/20395/files/46638cdef26fd25c290e6f63035ff5d2f1dc3bdb#diff-87bed125e3733038ed0310afc884fed6623fe7abe4f04bf7b8e0fd45ab73e79c

Reply:
ok, then.